### ISSUE TITLE
There may be a bug in create_reference_model

### ISSUE THREAD

**Issue by Sanster** [lvwerra/trl#244]
https://github.com/lvwerra/trl/blob/9b60207f0b1285799274a9ef7800209667dd3b94/trl/models/modeling_base.py#L394

Once shared_parameter = False, it will never be changed to True again.

```python
    shared_parameter = True
    for name, param in model.named_parameters():
        if pattern in name:
            shared_parameter = False
        if shared_parameter:
            shared_param_list.append(name)
        else:
            unshared_param_list.append(name)
```

**Comment by younesbelkada** [comment]
I think this is ok because `model.named_parameters()` will iterate over the parameters in the correct order. Assume your model has 4 layers: `transformer.h.0.xx`, `transformer.h.1.xx`, `transformer.h.2.xx`, `transformer.h.3.xx`, and `num_shared_layers=2`. In this case `pattern` would be equal to `transformer.h.2`

Inside the loop above, all the parameters before `transformer.h.2.xx` will remain untouched (i.e. `shared_parameter=True`), and all the parameters after `transformer.h.2.xx` (including it) will have `shared_parameters`, set to `False`. The same logic would totally work if `num_shared_layers=1` or `num_shared_layers=3`. 

Hence, there is no need  to change `shared_parameter` to `True` again. Does that makes sense to you?

**Comment by younesbelkada** [comment]
Thanks a lot!
To my understanding, 
```python
ref_param = param  # noqa
```
works as we're manipulating references here, also we do have some CI tests that effectively checks that the conversion has been performed correctly. I have now a doubt though with the fact that:
```python
name = shared_param_list[0]
print(id(ref_model.get_parameter(name)) == id(model.get_parameter(name)))
# False
```
Your proposed modification might makes sense, what do you think @lvwerra ?

**Comment by lvwerra** [comment]
Indeed, this should be manipulating the reference to the model and we test that it works. Not sure why `id` yields False but I am also not very familiar with it.
